src/java_bytecode/java_types.cpp

```diff
@@ -188,48 +188,53 @@ exprt java_bytecode_promotion(const exprt &expr)
     return typecast_exprt(expr, new_type);
 }
 
-/// Take a list of generic arguments and parse them into the generic type.
+/// Take a list of generic type variables and parse them into the generic type.
 /// \param generic_type [out]: The existing generic type to add the information
 ///   to
-/// \param parameters: The string representing the generic arguments for a
-///   signature. For example <TT;Ljava/lang/Foo;> (including wrapping angle
-///   brackets).
+/// \param type_variables: The string representing the generic type variables
+///   for a signature. For example `<TT;Ljava/lang/Foo;LList<LInteger;>;>`
+///   (including the wrapping angle brackets).
 /// \param class_name_prefix: The name of the class to use to prefix any found
 ///   generic types
 void add_generic_type_information(
   java_generic_typet &generic_type,
-  const std::string &parameters,
+  const std::string &type_variables,
   const std::string &class_name_prefix)
 {
-  PRECONDITION(parameters.size() >= 2);
-  PRECONDITION(parameters[0] == '<');
-  PRECONDITION(parameters[parameters.size() - 1] == '>');
+  PRECONDITION(type_variables.size() >= 2);
+  PRECONDITION(type_variables[0] == '<');
+  PRECONDITION(type_variables[type_variables.size() - 1] == '>');
 
-  // parse contained types, can be either type variables, starting with T
-  // or instantiated types
+  // Parse contained types, can be either type parameters (`TT;`)
+  // or instantiated types - either generic types (`LList<LInteger;>;`) or
+  // just references (`Ljava/lang/Foo;`)
   std::vector<typet> params =
-    parse_list_types(parameters, class_name_prefix, '<', '>');
+    parse_list_types(type_variables, class_name_prefix, '<', '>');
 
-  CHECK_RETURN(!params.empty()); // We should have at least one generic param
-
-  // take these types - they should either be java_generic_parameters, in which
-  // case they can be directly added to the generic_type
-  // otherwise they should be wrapped in a java_generic_inst_parametert
+  // We should have at least one generic type variable
+  CHECK_RETURN(!params.empty());
 
+  // Add the type variables to the generic type, and assign them the
+  // correct type
   std::transform(
     params.begin(),
     params.end(),
     std::back_inserter(generic_type.generic_type_variables()),
-    [](const typet &type) -> java_generic_parametert {
+    [](const typet &type) -> reference_typet
+    {
+      INVARIANT(
+        is_reference(type), "All generic type variables should be references");
       if(is_java_generic_parameter(type))
       {
         return to_java_generic_parameter(type);
       }
+      else if(is_java_generic_type(type))
```

Comment:
Good point. The whole `if` block is actually not needed here, `type` is of type `typet` and we just need to make sure it is a `reference_typet` but we don't need to cast it to any of its subtypes (it will be cast back to reference when returned). Will make the change.